File: web/src/api/packaging.ts
import api from "./client";

interface PaginatedResponse<T> {
  items: T[];
  total: number;
  limit: number;
  offset: number;
}

// ── Types ───────────────────────────────────────────────────

export interface PackagingStockItem {
  id: string;
  box_size_id: string | null;
  pallet_type_id: string | null;
  current_quantity: number;
  min_stock_level: number;
  name: string | null;
  weight_kg: number | null;
  cost_per_unit: number | null;
  packaging_type: string | null; // "box" | "pallet"
  created_at: string;
  updated_at: string;
}

export interface PackagingMovement {
  id: string;
  stock_id: string;
  movement_type: string; // "receipt" | "consumption" | "adjustment"
  quantity: number;
  cost_per_unit: number | null;
  reference_type: string | null;
  reference_id: string | null;
  notes: string | null;
  recorded_by: string | null;
  recorded_at: string;
}

// ── Receipt payload ─────────────────────────────────────────

export interface PackagingReceiptPayload {
  box_size_id?: string;
  pallet_type_id?: string;
  quantity: number;
  cost_per_unit?: number;
  notes?: string;
}

// ── API calls ───────────────────────────────────────────────

export async function getPackagingStock(): Promise<PackagingStockItem[]> {
  const { data } = await api.get<PackagingStockItem[]>("/packaging/stock");
  return data;
}

export async function receivePackaging(
  payload: PackagingReceiptPayload
): Promise<PackagingStockItem> {
  const { data } = await api.post<PackagingStockItem>("/packaging/receipt", payload);
  return data;
}

export async function updateMinStock(
  stockId: string,
  min_stock_level: number
): Promise<PackagingStockItem> {
  const { data } = await api.patch<PackagingStockItem>(
    `/packaging/stock/${stockId}/min`,
    { min_stock_level }
  );
  return data;
}

export async function adjustStock(
  stock_id: string,
  quantity: number,
  notes?: string
): Promise<PackagingStockItem> {
  const { data } = await api.post<PackagingStockItem>("/packaging/adjustment", {
    stock_id,
    quantity,
    notes,
  });
  return data;
}

export async function listMovements(
  params?: Record<string, string>
): Promise<PackagingMovement[]> {
  const { data } = await api.get<PaginatedResponse<PackagingMovement>>(
    "/packaging/movements",
    { params }
  );
  return data.items;
}

File: backend/app/schemas/packaging.py
"""Pydantic schemas for packaging stock management."""

from datetime import datetime

from pydantic import BaseModel, Field


# ── Stock overview ──────────────────────────────────────────

class PackagingStockOut(BaseModel):
    id: str
    box_size_id: str | None = None
    pallet_type_id: str | None = None
    current_quantity: int
    min_stock_level: int

    # Resolved names from relationships
    name: str | None = None
    weight_kg: float | None = None
    cost_per_unit: float | None = None
    packaging_type: str | None = None  # "box" or "pallet"

    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── Receipt (stock in) ─────────────────────────────────────

class PackagingReceiptRequest(BaseModel):
    """Import / receive packaging stock."""
    box_size_id: str | None = None
    pallet_type_id: str | None = None
    quantity: int = Field(..., ge=1)
    cost_per_unit: float | None = None
    notes: str | None = None


# ── Adjustment ──────────────────────────────────────────────

class PackagingAdjustmentRequest(BaseModel):
    """Manual stock correction (positive or negative)."""
    stock_id: str
    quantity: int  # positive to add, negative to subtract
    notes: str | None = None


# ── Min stock level ─────────────────────────────────────────

class UpdateMinStockRequest(BaseModel):
    min_stock_level: int = Field(..., ge=0)


# ── Movement history ────────────────────────────────────────

class PackagingMovementOut(BaseModel):
    id: str
    stock_id: str
    movement_type: str
    quantity: int
    cost_per_unit: float | None = None
    reference_type: str | None = None
    reference_id: str | None = None
    notes: str | None = None
    recorded_by: str | None = None
    recorded_at: datetime

    model_config = {"from_attributes": True}
